docs/proposals/scheduling/20230707-eviction-arbitrate-mechanism-in-descheduler.md

```diff
@@ -0,0 +1,256 @@
+- **Arbitration Process**: This process periodically arbitrates the PodMigrationJobs in the arbitration collection and places the PodMigrationJobs selected by the arbitration into the rate limit queue.
+- **Migration Process**: This process will read `PodMigrationJob` from the rate limit queue and call `doMigrat`e to handle the `PodMigrationJob`.
+
+###### Controller Reconcile Process
+
+1. Get the `PodMigrationJob`.
+2. If PodMigrationJob's phase is empty or `pending`, put it to the arbitration collection.
+3. Else do other things.
+
+###### Arbitration Process
+
+1. Sort the elements in the arbitration collection to generate a slice.
+2. Use a map to record the sorted positions of each `PodMigrationJob` in the slice, with key being the `PodMigrationJob` and value being the position.
+3. Call different `GroupFilter` functions in sequence for group and filter operations to update the slice.
+4. Resort the elements in the slice using the previous map.
+5. Place the first n elements of slice to the rate limit queue and remove these elements from the arbitration collection.
+
+###### Migration Process
+
+1. For each element in the rate limit queue, call `doMigrate` to process the PodMigrationJob.
+
+##### Arbitration Mechanism
+
+The Arbitration Mechanism works after the descheduler and mainly manages the PodMigrationJobs generated by the descheduler. It decides which PodMigrationJob to execute, to provide a guarantee for the stability of the descheduler and migration, and improve the stability of the system and applications.
+
+It arbitrates the PodMigrationJobs in the arbitration collection and places the PodMigrationJobs already passed the arbitration into the rate limit queue.
+
+###### Sort PodMigrationJob
+
+![image](/docs/images/arbitration-mechanism-sort-design.svg)
+
+- Using the **stable sorting** method, sort PodMigrationJobs in the following order:
+  - The time interval between the start of migration and the current, the smaller the interval, the higher the ranking.
+  - The Pod priority of PodMigrationJob, the lower the priority, the higher the ranking.
+  - If some Pods in the job containing PodMigrationJob's Pod is being migrated, the PodMigrationJob's ranking is higher.
+  - The higher the migration priority, the higher the ranking.
+  - BE > LS > LSR = LSE
+- Use a map to record the sorted ranking (key is PodMigrationJob, value is position)
+- Set all `PodMigrationJob` ranking under the same job in the map to the highest one of them (Try to migrate Pods from the same job simultaneously as much as possible).
+
+The definition of the type `SortFn` is as follows.
+
+~~~ go
+// SortFn stably sorts PodMigrationJobs slice based on a certain strategy. Users 
+// can implement different SortFn according to their needs.
+type SortFn func(jobs []*v1alpha1.PodMigrationJob, client client.Client) []*v1alpha1.PodMigrationJob
+~~~
+
+###### GroupFilter PodMigrationJob
```

Comment:
In the initial conception, the arbitration mechanism is actually a shaping process. It has a large overlap with the existing rate-limiting mechanism in the migration controller. However, the existing rate-limiting mechanism is passive, and arbitration can refer to more dimensional information.

When a MigrationJob is created, MigrationController first delivers the MigrationJob to the arbitration queue, then the arbitrator performs rate-limiting and shaping, and then delivers it to the Reconcile queue. From a structural point of view, this is very clear, so I am wondering whether the arbitration mechanism should be used to replace the existing rate-limiting mechanism, otherwise it is very strange that descheduler has two sets of rate-limiting methods.

WDYT?

Reply:
@ZiMengSheng  PTAL.

Reply:
The arbitration queue, I think, is not a rate-limiting queue but a collection to keep the MigrationJobs In a period of time to be arbitrated. After arbitration, the job will be placed into the reconcile queue, which is the true rate-limiting queue.

Reply:
> 

Yes ！